keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,737 @@
+should keep this high-level with a focus on what signals will be looked at to
+determine graduation.
+
+Consider the following in developing the graduation criteria for this enhancement:
+- [Maturity levels (`alpha`, `beta`, `stable`)][maturity-levels]
+- [Deprecation policy][deprecation-policy]
+
+Clearly define what graduation means by either linking to the [API doc
+definition](https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-versioning),
+or by redefining what graduation means.
+
+In general, we try to use the same stages (alpha, beta, GA), regardless how the
+functionality is accessed.
+
+[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
+[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+
+Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].
+
+#### Alpha -> Beta Graduation
+
+- Gather feedback from developers and surveys
+- Complete features A, B, C
+- Tests are in Testgrid and linked in KEP
+
+#### Beta -> GA Graduation
+
+- N examples of real world usage
+- N installs
+- More rigorous forms of testing e.g., downgrade tests and scalability tests
+- Allowing time for feedback
+
+**Note:** Generally we also wait at least 2 releases between beta and
+GA/stable, since there's no opportunity for user feedback, or even bug reports,
+in back-to-back releases.
+
+**For non-optional features moving to GA, the graduation criteria must include [conformance tests].**
+
+[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+-->
+
+TODO
+
+### Upgrade / Downgrade Strategy
+
+#### `--node-ip` deprecation
+
+This can happen in the usual way for deprecating command-line
+arguments.
+
+#### IPv6 `Node.Status.Addresses` Gotchas
+
+The proposed behavior has one upgrade caveat:
+
+  - If you currently have a single-stack IPv4 cluster,
+  - _and_ aren't passing any `--node-ip` value to kubelet,
+  - _and_ have software in/around the cluster that will fail if it
+    sees an IPv6 address in `Node.Status.Addresses` even when that
+    address comes after the Primary Node IP.
+  - _and_ are running on a cloud that supports IPv6
+  - _and_ are running on hosts that have IPv6 addresses assigned to
+    them even though you weren't using them (at least for Kubernetes),
+  - then you will need to update your kubelet configuration to include
+    `--node-ips ipv4` to force kubelet to be single-stack-IPv4-only in
+    the future.
+
+This should hopefully be uncommon enough to be essentially
+non-existent. And also, this does not _technically_ constitute an API
+break, since `Node.Status.Addresses` was always theoretically
+dual-stack.
+
+The alternative to this would be to make it so that if you pass no
+`--node-ips` you always get a single-stack IPv4 cluster, even in a
+dual-stack environment, which seems non-future-friendly.
```

Comment:
It is somewhat compatible with the rest of dual-stack (very conservative defaults).  That said, I kind of agree that NodeAddresses carrying both seems probably reasonable.